### PR TITLE
Solves Reduce the Threads "Dir Watcher" count #3129

### PR DESCRIPTION
**Pull request by adimova**
Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>

### CODE REVIEW

File: bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java [hunk 1]
```diff
@@ -69,9 +66,23 @@
  *
  * @author Kai Kreuzer - Initial contribution and API
  * @author Petar Valchev - Added sort by modification time, when configuration files are read
+ * @author Ana Dimova - reduce to a single watch thread for all class instances
  */
 public class ConfigDispatcher extends AbstractWatchService {
 
+    private static String getPathToWatch() {
+        String progArg = System.getProperty(SERVICEDIR_PROG_ARGUMENT);
```
Comment (kaikreuzer) [inline bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java]: Could we actually replace this automation specific "servicedir" by `ConfigConstants.getConfigFolder()`, which is the ESH global definition of the configuration directory?

Comment (adimova) [inline bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java]: probably we could, but I actually do not understand how should I change it

Comment (adimova) [inline bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java]: Is this what you mean?
private static String getPathToWatch() {
            return ConfigConstants.getConfigFolder() + File.separator + SERVICES_FOLDER;
}
or
private static String getPathToWatch() {
            return ConfigConstants.getConfigFolder();
}

Comment (kaikreuzer) [inline bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java]: Err, sorry, forget my comment - I actually didn't notice that this is within the ConfigDispatcher class, I somehow thought it was related to automation... Never mind!

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/WatchQueueReader.java [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+            boolean watchSubDirectories) {
+        try {
+            if (watchSubDirectories) {
+                // walk through all folders and follow symlinks
+                registerWithSubDirectories(watchService, toWatch);
+            } else {
+                registerDirectoryInternal(watchService, watchService.getWatchEventKinds(toWatch), toWatch);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/WatchQueueReader.java]: I assume this should be deleted

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/WatchQueueReader.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+    private synchronized void registerDirectoryInternal(AbstractWatchService service, Kind<?>[] kinds, Path directory) {
+        if (watchService == null) {
+            try {
+                watchService = FileSystems.getDefault().newWatchService();
+                qr = new Thread(this, "Dir Watcher");
+                qr.start();
+            } catch (IOException e) {
+                e.printStackTrace();
+                logger.error("Cannot activate watch service", e.getMessage());
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/service/WatchQueueReader.java]: '{}' is missing in the string

**Pull request merged**